python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -373,17 +378,21 @@ def next_version_record(
         after_cursor: Optional[int],
         data_version: Optional[DataVersion],
     ) -> Optional["EventLogRecord"]:
-        for record in self.instance.get_event_records(
-            EventRecordsFilter(
-                event_type=DagsterEventType.ASSET_OBSERVATION,
-                asset_key=asset_key,
-                after_cursor=after_cursor,
-            ),
-            ascending=True,
-        ):
-            record_version = extract_data_version_from_entry(record.event_log_entry)
-            if record_version is not None and record_version != data_version:
-                return record
+        has_more = True
+        cursor = None
+        while has_more:
```

Comment:
i guess the way that I would do it, just to avoid bugs is to name the helper fn something like:
```
def fetch_materializations_in_a_loop_which_is_probably_bad_design(instance, asset_key, ...):
    ...
```

Reply:
What in particular should feel bad? Looking at older records than the most recent record?

Reply:
One case I'm thinking about here is asset checks that do anomaly detection and need to compare present values to multiple historical values

Reply:
I think fetching an unknown number of records in a single call is dangerous.  It hides the fact that this single db call is creating a lot of load on the DB.

Reply:
What exactly do you mean by "unknown"? In the anomaly detection case, the user would specify a number of records they want. But they would still need to write this loop, right?